src/platform/nxp/k32w/k32w0/RamStorage.cpp

```diff
@@ -0,0 +1,201 @@
+#endif
+
+static rsError ramStorageAdd(ramBufferDescriptor * pBuffer, uint16_t aKey, const uint8_t * aValue, uint16_t aValueLength)
+{
+    rsError error;
+    struct settingsBlock * currentBlock;
+    const uint16_t newBlockLength = sizeof(struct settingsBlock) + aValueLength;
+
+    if (pBuffer->ramBufferLen + newBlockLength <= pBuffer->ramBufferMaxLen)
+    {
+        currentBlock         = (struct settingsBlock *) &pBuffer->pRamBuffer[pBuffer->ramBufferLen];
+        currentBlock->key    = aKey;
+        currentBlock->length = aValueLength;
+
+        memcpy(&pBuffer->pRamBuffer[pBuffer->ramBufferLen + sizeof(struct settingsBlock)], aValue, aValueLength);
+        pBuffer->ramBufferLen += newBlockLength;
+
+        error = RS_ERROR_NONE;
+    }
+    else
+    {
+        error = RS_ERROR_NO_BUFS;
+    }
+
+    RAM_STORAGE_PRINTF("key = %d lengthWriten = %d err = %d", aKey, aValueLength, error);
+
+    return error;
+}
+
+rsError ramStorageGet(const ramBufferDescriptor * pBuffer, uint16_t aKey, int aIndex, uint8_t * aValue, uint16_t * aValueLength)
+{
+    uint16_t i           = 0;
+    uint16_t valueLength = 0;
+    uint16_t readLength;
+    int currentIndex = 0;
+    const struct settingsBlock * currentBlock;
+    rsError error = RS_ERROR_NOT_FOUND;
+
+    while (i < pBuffer->ramBufferLen)
+    {
+        currentBlock = (struct settingsBlock *) &pBuffer->pRamBuffer[i];
+
+        if (aKey == currentBlock->key)
+        {
+            if (currentIndex == aIndex)
+            {
+                readLength = currentBlock->length;
+
+                // Perform read only if an input buffer was passed in
+                if (aValue != NULL && aValueLength != NULL)
+                {
+                    // Adjust read length if input buffer size is smaller
+                    if (readLength > *aValueLength)
+                    {
+                        readLength = *aValueLength;
+                    }
+
+                    memcpy(aValue, &pBuffer->pRamBuffer[i + sizeof(struct settingsBlock)], readLength);
+                }
+
+                valueLength = currentBlock->length;
+                error       = RS_ERROR_NONE;
+                break;
+            }
+
+            currentIndex++;
+        }
+
+        i += sizeof(struct settingsBlock) + currentBlock->length;
+    }
+
+    if (aValueLength != NULL)
+    {
+        *aValueLength = valueLength;
+    }
+
+    RAM_STORAGE_PRINTF("key = %d err = %d", aKey, error);
+
+    return error;
+}
+
+rsError ramStorageSet(ramBufferDescriptor * pBuffer, uint16_t aKey, const uint8_t * aValue, uint16_t aValueLength)
+{
+    uint16_t i = 0;
+    uint16_t currentBlockLength;
+    uint16_t nextBlockStart;
+    const struct settingsBlock * currentBlock;
+
+    // Delete all entries of aKey
+    while (i < pBuffer->ramBufferLen)
+    {
+        currentBlock       = (struct settingsBlock *) &pBuffer->pRamBuffer[i];
+        currentBlockLength = sizeof(struct settingsBlock) + currentBlock->length;
```

Comment:
This is dangerous code. If you wrote/stored raw data, please use `memcpy` to copy in/out of buffers, rather than type punning. Otherwise, make *very clear* in your code that everything has to be word aligned, since this code is doing pointer aliasing and may violate alignment rules. It probably works b/c you are on M3/M4, but this exact code could fail on Cortex-M0.